src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -0,0 +1,222 @@
+            //
+            // distance is how many iterations is the entry from it ideal position.
+            // we use that for preemption.
+            //
+            // versionNum is a monotonicaly increasing numerical tag.
+            // Writer "claims" entry by atomically incrementing the tag. Thus odd number indicates an entry in progress.
+            // Upon completion of adding an entry the tag is incremented again making it even. Even number indicates a complete entry.
+            //
+            // Readers will read the version twice before and after retrieving the entry.
+            // To have a usable entry both reads must yield the same even version.
+            //
+            internal int  _version;
+            internal nint _source;
+            // pointers have unused lower bits due to alignment, we use one for the result
+            internal nint _targetAndResult;
+        };
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int KeyToBucket(int[] table, nint source, nint target)
+        {
+            // upper bits of addresses do not vary much, so to reduce loss due to cancelling out,
+            // we do `rotl(source, <half-size>) ^ target` for mixing inputs.
+            // then we use fibonacci hashing to reduce the value to desired size.
+
+#if BIT64
+            ulong hash = (((ulong)source << 32) | ((ulong)source >> 32)) ^ (ulong)target;
+            return (int)((hash * 11400714819323198485ul) >> HashShift(table));
+#else
+            uint hash = (((uint)source >> 16) | ((uint)source << 16)) ^ (uint)target;
+            return (int)((hash * 2654435769ul) >> HashShift(table));
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ref int AuxData(int[] table)
+        {
+            // element 0 is used for embedded aux data
+            return ref Unsafe.As<byte, int>(ref table.GetRawSzArrayData());
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ref CastCacheEntry Element(int[] table, int index)
+        {
+            // element 0 is used for embedded aux data, skip it
+            return ref Unsafe.Add(ref Unsafe.As<int, CastCacheEntry>(ref AuxData(table)), index + 1);
+        }
+
+        // TableMask is "size - 1"
+        // we need that more often that we need size
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int TableMask(int[] table)
+        {
+            return AuxData(table);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int HashShift(int[] table)
+        {
+            return Unsafe.Add(ref AuxData(table), 1);
+        }
+
+        private enum CastResult
+        {
+            CannotCast = 0,
+            CanCast = 1,
+            MaybeCast = 2
+        }
+
+        // NOTE!!
+        // This is a copy of C++ implementation in CastCache.cpp
```

Comment:
More likely `castcache.h`.

Reply:
Nit: The implementation is in the .cpp file, but the VM filename is in lower case (castcache.cpp).